File: doubly_linkedlist.py
class Node:
    def __init__(self, val):
        self.val = val
        self.prev = None
        self.nex = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.sz = 0

    def insert_node(self, anode):
        """
        insert a node to the head of the list, O(1) time complexity
        """
        if self.head:
            anode.nex = self.head
            self.head.prev = anode
        self.head = anode
        self.sz += 1

    def delete_node(self, anode):
        """
        delete a node at any place if exists in the list
        """
        if anode.prev:
            anode.prev.nex = anode.nex
            self.sz -= 1
        if self.head == anode:
            self.head = self.head.nex
            self.sz -= 1
        if anode.nex:
            anode.nex.prev = anode.prev

    def traversal(self):
        dummy = self.head
        while dummy:
            print(dummy.val)
            dummy = dummy.nex

    def size(self):
        return self.sz

    def top(self):
        if self.head:
            return self.head.val

    def search(self, key):
        dummy = self.head
        while dummy:
            if dummy.val == key:
                return True
        return False



if __name__ == "__main__":
    node1 = Node(3)
    node2 = Node(12)
    node3 = Node(99)
    dlist = DoublyLinkedList()
    dlist.delete_node(node3)
    dlist.insert_node(node1)
    dlist.insert_node(node2)
    dlist.insert_node(node3)
    dlist.delete_node(node1)
    dlist.traversal()
    print(dlist.top())

        

File: doubly_linkedlist.cpp
#include <iostream>
#include <climits>

struct Node
{
    int val;
    Node* prev = nullptr;
    Node* next = nullptr;
};

class DoublyLinkedList
{
    private:
        Node* head = nullptr;
        int sz = 0;
    public:
        DoublyLinkedList(){};
        ~DoublyLinkedList(){};
        void insert_val(int val)
        {
            Node* new_p = new Node;
            new_p->val = val;
            new_p->next = head;
            if(head != nullptr)
            {
                head->prev = new_p;
            }
            head = new_p;
            sz++;
        }
        void delete_val(int val)
        {
            Node* new_p = new Node;
            new_p->val = 0;
            new_p->next = head;
            Node* final_list = new_p;
            while(new_p->next != nullptr)
            {
                if(new_p->next->val == val)
                {
                    new_p->next = new_p->next->next;
                    //now new_p->next is actually old new_p->next->next
                    if(new_p->next != nullptr)
                        new_p->next->prev = new_p;
                    sz--;
                    break;
                }
                new_p = new_p -> next;
            }
            head = final_list->next;
        }
        void traversal()
        {
            Node* dummy = head;
            while(dummy != nullptr)
            {
                std::cout << dummy->val << std::endl;
                dummy = dummy->next;
            }
        }
        int top()
        {
            if(head != nullptr)
                return head->val;
            return INT_MAX;
        }
        int size()
        {
            return sz;
        }
};

int main()
{

    DoublyLinkedList dlist;
    dlist.delete_val(12);
    dlist.insert_val(3);
    dlist.insert_val(9);
    dlist.insert_val(45);
    std::cout << dlist.size() << std::endl;
    dlist.delete_val(3);
    std::cout << dlist.top() << std::endl;
    dlist.traversal();
    return 0;
}